### PR TITLE
feat: add DuckDuckGo search tool

### PR DESCRIPTION
**Pull request by leehuwuj**
## Summary by CodeRabbit

- **New Features**
  - Introduced DuckDuckGo search functionality in Python and TypeScript.

- **Dependencies**
  - Added `duck-duck-scrape` dependency to projects using Express and Next.js.

### CODE REVIEW

File: templates/components/engines/typescript/agent/tools/duckduckgo.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  },
+};
+
+type DuckDuckGoSearchResult = {
+  title: string;
+  description: string;
+  url: string;
+};
+
+export class DuckDuckGoSearchTool implements BaseTool<DuckDuckGoParameter> {
```
Comment (leehuwuj) [inline templates/components/engines/typescript/agent/tools/duckduckgo.ts:41]: The DuckDuckGo in python have two functions:
+ instant_search: Get an instant answer from DDG.
+ search: Get the search results.

But the library we're using does not support instant_search, this is a different tool in two languages.

File: templates/components/engines/typescript/agent/tools/index.ts [hunk 2]
```diff
@@ -35,6 +36,9 @@ const toolFactory: Record<string, ToolCreator> = {
     );
     return await openAPIActionTool.toToolFunctions();
   },
+  "duckduckgo.DuckDuckGoSearchToolSpec": async (config: unknown) => {
```
Comment (leehuwuj) [inline templates/components/engines/typescript/agent/tools/index.ts:39]: In python, it's a llamahub tool but in TS it's a local tool. Do you have any suggestions?

File: templates/components/engines/typescript/agent/tools/duckduckgo.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    return searchResults.results.map((result) => {
+      return {
+        title: result.title,
+        description: result.description,
+        url: result.url,
+      } as DuckDuckGoSearchResult;
+    });
+  }
+}
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/components/engines/typescript/agent/tools/duckduckgo.ts:61]: @leehuwuj this remark looks correct

File: templates/components/engines/typescript/agent/tools/duckduckgo.ts (hunk 1 above)
Comment (marcusschiesser) [reply templates/components/engines/typescript/agent/tools/duckduckgo.ts:41]: got it should be a comment

File: templates/components/engines/typescript/agent/tools/index.ts (hunk 2 above)
Comment (marcusschiesser) [reply templates/components/engines/typescript/agent/tools/index.ts:39]: just a comment would be good

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**